evm-adapters/src/sputnik/cheatcodes/debugger.rs

```diff
@@ -0,0 +1,403 @@
+use sputnik::{
+    Capture, ExitReason, ExitSucceed, Handler, Machine, Memory, Opcode, Resolve, Runtime,
+};
+
+use ethers::types::{Address, H256};
+
+use std::{borrow::Cow, fmt::Display, rc::Rc};
+/// EVM runtime.
+///
+/// The runtime wraps an EVM `Machine` with support of return data and context.
+pub struct ForgeRuntime<'b, 'config> {
+    pub inner: &'b mut Runtime<'config>,
+    pub code: Rc<Vec<u8>>,
+}
+
+#[derive(Debug, Clone)]
+/// An arena of `DebugNode`s
+pub struct DebugArena {
+    /// The arena of nodes
+    pub arena: Vec<DebugNode>,
+    /// The entry index, denoting the first node's index in the arena
+    pub entry: usize,
+}
+
+impl Default for DebugArena {
+    fn default() -> Self {
+        DebugArena { arena: vec![Default::default()], entry: 0 }
+    }
+}
+
+impl DebugArena {
+    /// Pushes a new debug node into the arena
+    pub fn push_node(&mut self, entry: usize, mut new_node: DebugNode) {
+        match new_node.depth {
+            // The entry node, just update it
+            0 => {
+                self.arena[entry] = new_node;
+            }
+            // we found the parent node, add the new node as a child
+            _ if self.arena[entry].depth == new_node.depth - 1 => {
+                new_node.idx = self.arena.len();
+                new_node.location = self.arena[entry].children.len();
+                self.arena[entry].children.push(new_node.idx);
+                self.arena.push(new_node);
+            }
+            // we haven't found the parent node, go deeper
+            _ => self.push_node(
+                *self.arena[entry].children.last().expect("Disconnected debug node"),
+                new_node,
+            ),
+        }
+    }
+
+    pub fn flatten(&self, entry: usize, flattened: &mut Vec<(Address, Vec<DebugStep>)>) {
```

Comment:
maybe make this a bit clearer, like

```suggestion
    /// Recursively traverses the tree of debug step nodes and flattens it into a 
    /// vector where each element contains
    /// 1. the address of the contract being executed
    /// 2. a vector of all the debug steps along that contract's execution path.
    ///  
    /// This then makes it easy to pretty print the execution steps.
    pub fn flatten(&self, entry: usize, flattened: &mut Vec<(Address, Vec<DebugStep>)>) {

```